ipq40xx: add support for MikroTik SXTsq 5 ac

rogerpueyo opened the pull request:
**Note: this pull request is largely based on @robimarko's [support for the MikroTik hAP ac² PR](https://github.com/openwrt/openwrt/pull/3037), ~~and depends on several commits there being merged upstream.~~**

**Edit: [support for MikroTik hAP ac²](https://github.com/openwrt/openwrt/pull/3037) is merged upstream.** :+1: 

This commit adds support for the MikroTik SXTsq 5 ac (RBSXTsqG-5acD), an outdoor 802.11ac wireless CPE with one 10/100/1000 Mbps Ethernet port.

Specifications:
 - SoC: Qualcomm Atheros IPQ4018
 - RAM: 256 MB
 - Storage: 16 MB NOR
 - Wireless: IPQ4018 (SoC) 802.11a/n/ac 2x2:2, 16 dBi antennae
 - Ethernet: IPQ4018 (SoC) 1x 10/100/1000 port, 10-28 Vdc PoE in
 - 7 user-controllable LEDs
  · 1x power (blue)
  · 1x user (green)
  · 5x rssi (green)
 - 1 Ethernet LED (green)

Note:
 Serial UART is probably available on the board, but it has not been tested.

Flashing:
 Boot via TFTP the initramfs image. Then, upload a sysupgrade image via SSH and flash it normally. More info at the "Common procedures for MikroTik products" page https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>

adrianschmutzler commented:
> CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y

IIRC, this will break sysupgrade compatibility for SPI-based devices?

robimarko commented:
Should not as its limited to under 4MB NOR flash chips

markbirss commented:
Sysupgrade is not successful for me?

rogerpueyo commented:
@markbirss ,

Sorry, I am afraid that in the last rebase/push I ended up removing a few lines from the DTS that are actually needed:

```
  serial_pins: serial_pinmux {
  		mux {
  			pins = "gpio60", "gpio61";
  			function = "blsp_uart0";
  			bias-disable;
  		};
  	};

[...]

&blsp1_uart1 {
	status = "okay";

	pinctrl-0 = <&serial_pins>;
	pinctrl-names = "default";
};
```

Could you please try with the latest commit dc49391ab8e55d425be98ed0a1b5678672174645? You should now see the "user" LED blinking now during sysupgrade.

markbirss commented:
@rogerpueyo

Ok sure, i will have a look, thank you for you reply

markbirss commented:
@rogerpueyo

yes, thank you those changes and commit dc49391, sysupgrade working!!

markbirss commented:
@rogerpueyo

the only issue i found was that i needed to enable

`uci set dhcp.lan.force='1'`

or i would not get a dhcp lease 

Im now running mesh with LibreMesh (batman-adv)

rogerpueyo commented:
@markbirss 

Cool, thanks for testing it.



> the only issue i found was that i needed to enable
> `uci set dhcp.lan.force='1'`

I would say this doesn't look to me like a device-specific issue. Is it happening with plain OpenWrt or with LiMe? I'm getting proper DHCP leases on the LAN port, both cable and wifi.

hodgesds commented:
I got this working as well, thanks so much for the work!

markbirss commented:
I used dnsmasq dhcp command line server to send the initramfs-kernel, this
could have caused my issue at the time

Openwrt detected an already running dhcp


On Mon, 15 Jun 2020, 03:10 Daniel Hodges, <notifications@github.com> wrote:

> I got this working as well, thanks so much for the work!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/3108#issuecomment-643852851>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFKZ2J462ZJA6HYSLSF4E4LRWVYHDANCNFSM4N5DA5KQ>
> .
>

f00b4r0 commented:
I just noticed this PR.

> > root@OpenWrt:/# cat /proc/mtd
> > dev:    size   erasesize  name
> > mtd0: 01000000 00001000 "partitions"
> > mtd1: 0000e000 00001000 "RouterBoot"
> > mtd2: 00001000 00001000 "hard_config"
> > mtd3: 00007dab 00001000 "dtb_config"
> > mtd4: 00001000 00001000 "soft_config"
> > mtd5: 00f00000 00001000 "firmware"

This shows exactly that 4K_LIMITS is currently broken, and @adrianschmutzler's concerns are valid.

This also uses the frankly awful "loader" from #3037 and the [same remarks apply](https://github.com/openwrt/openwrt/pull/3037#issuecomment-633104108)

Given the discoveries made in #3026, I think another way is open, as [laid out here](https://github.com/openwrt/openwrt/pull/3037#issuecomment-643651573).

My current 2c :)

rogerpueyo commented:
> This also uses the frankly awful "loader" from #3037 and the [same remarks apply](https://github.com/openwrt/openwrt/pull/3037#issuecomment-633104108)

My commit actually depends on #3037 (first line of the PR message :wink:), so I assume it must be finished and merged upstream first.

> This shows exactly that 4K_LIMITS is currently broken, and @adrianschmutzler's concerns are valid.

Indeed, writing to the NOR flash is significantly slow.

> Given the discoveries made in #3026, I think another way is open, as laid out here.

Yes, I had read that but I am afraid I don't know how to help on that (other than testing on my device). :disappointed:

f00b4r0 commented:
My current understanding is that since 4.19, both MTD_SPI_NOR_USE_4K_SECTORS_LIMIT and MTD_SPI_NOR_USE_4K_SECTORS are partially or entirely broken.

The mtd subsystem switched in 4.17 from an asynchronous API to a synchronous one, removing in particular some callbacks that were used in the implementation of the above.

The net result is that the resulting code does not work as intended/at all (this is a **treewide** issue), and I'm unfortunately not able to propose a fix for I don't understand this subsystem and the original code (submitted by @nbd168 iirc) well enough. Another pair of eyes is needed I'm afraid.

SvenRoederer commented:
As the is some confusion on the MikroTik productnames, I opened #3270, where we can probably agree on a guideline.

f00b4r0 commented:
> My current understanding is that since 4.19, both MTD_SPI_NOR_USE_4K_SECTORS_LIMIT and MTD_SPI_NOR_USE_4K_SECTORS are partially or entirely broken.
> 
> The mtd subsystem switched in 4.17 from an asynchronous API to a synchronous one, removing in particular some callbacks that were used in the implementation of the above.
> 
> The net result is that the resulting code does not work as intended/at all (this is a **treewide** issue), and I'm unfortunately not able to propose a fix for I don't understand this subsystem and the original code (submitted by @nbd168 iirc) well enough. Another pair of eyes is needed I'm afraid.

FYI @john-tho is looking into this in #3103

f00b4r0 commented on target/linux/ipq40xx/files-5.4/arch/arm/boot/dts/qcom-ipq4018-hap-ac2.dts:
```diff
@@ -0,0 +1,236 @@
+		spi-max-frequency = <24000000>;
+
+		partitions {
+			compatible = "mikrotik,routerboot-partitions";
```
Please don't do this. As I [explained before](https://github.com/openwrt/openwrt/pull/3037/files#r465891819) this is not how it's meant to be. Please re-use the [same construct we used for ath79](https://github.com/openwrt/openwrt/blob/3d167ed805a9300e60f8edcccfd740786b67e671/target/linux/ath79/dts/qca9556_mikrotik_routerboard-wap-g-5hact2hnd.dts#L71).

rogerpueyo replied:
Nice, I'm giving it a try.

rogerpueyo replied:
Copied hap ac²'s `partitions` node.

f00b4r0 commented:
@rogerpueyo I think you really didn't want to push those hundreds of commits. Failed rebase? 😛

rogerpueyo commented:
> @rogerpueyo I think you really didn't want to push those hundreds of commits. Failed rebase? stuck_out_tongue

Definitely... :facepalm:

rogerpueyo commented:
Well, I'm not sure what happened. Once the hAP ac² is finally upstreamed I'll rebase from master and apply my patch cleanly. :disappointed:

timprepscius commented:
Greetings gods of openwrt!   Is this a branch which I can possibly test out?  I would like to use openwrt on a sxtsq-5, and am considering buying a second to test.

hodgesds commented:
Yeah, in theory you can build this branch and run it. There's a few threads in the openwrt forums as well with some info. I might even have a kernel built stored somewhere.

timprepscius commented:
Ok, I will order the router and let you know, will be back probably in 2 weeks.

rogerpueyo committed:
ipq40xx: add support for MikroTik SXTsq 5 ac

This commit adds support for the MikroTik SXTsq 5 ac (RBSXTsqG-5acD),
an outdoor 802.11ac wireless CPE with one 10/100/1000 Mbps Ethernet
port.

Specifications:
 - SoC: Qualcomm Atheros IPQ4018
 - RAM: 256 MB
 - Storage: 16 MB NOR
 - Wireless: IPQ4018 (SoC) 802.11a/n/ac 2x2:2, 16 dBi antennae
 - Ethernet: IPQ4018 (SoC) 1x 10/100/1000 port, 10-28 Vdc PoE in
 - 1x Ethernet LED (green)
 - 7x user-controllable LEDs
  · 1x power (blue)
  · 1x user (green)
  · 5x rssi (green)

Note:
 Serial UART is probably available on the board, but it has not been
 tested.

Flashing:
 Boot via TFTP the initramfs image. Then, upload a sysupgrade image
 via SSH and flash it normally. More info at the "Common procedures
 for MikroTik products" page https://openwrt.org/toh/mikrotik/common.

Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>

rogerpueyo commented:
> Ok, I will order the router and let you know, will be back probably in 2 weeks.

I updated the PR, including the latest commits from https://github.com/robimarko/openwrt/tree/hAP-ac2-cleanup

f00b4r0 commented:
@rogerpueyo I'd wait for #3037 to be fully reviewed and merged before force pushing a whole bunch of patches to this PR, if possible: let's avoid duplicating review efforts. My 2c :)

rogerpueyo commented:
@f00b4r0 Sure! I assume this PR will only be ready once the hAP one is upstreamed. :+1:

spolack commented:
I tested SXTsq 5 ac built using your branch - Works for me!

rogerpueyo commented:
Hi Simon,

Thanks for testing it! This patch depends on the hAP ac² so, once that 
one is merged upstream, I'll refresh this one and kindly ask you to test 
it again.

Cheers!

El 21/11/20 a les 17:54, Simon Polack ha escrit:
>
> I tested SXTsq 5 ac built using your branch - Works for me!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/openwrt/openwrt/pull/3108#issuecomment-731604669>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABN3GA6KOOXS5WZAAB5QXNLSQ7WDXANCNFSM4N5DA5KQ>.
>

spolack commented:
I rebased on current branch from robimako and i can tell, both initramfs and sysupgrade images are booting and almost fully working ...

Except sysupgrade from the rebased images seems to be broken and results in a boot loop. Do you have any idea what can i do about it?

Best
Simon

f00b4r0 commented:
Depends on how you "rebased". the hap-ac2 PR and this PR have conflicting changes (4K_SECTORS, pre_upgrade()) which can explain your symptom.

robimarko commented:
@f00b4r0 I think that I caused the regression when removing the `platform_pre_upgrade` call and integrating the rootfs erase into the upgrade call.

For whatever reason now mtd erase is not executed once running initramfs.

f00b4r0 commented:
@robimarko in fact you didn't, it's the whole removal of platform_pre_upgrade() that seems to have broken this for us. I just had a chat with @damex on IRC, he's experiencing the exact same problem on ramips devices.

I understand @rmilecki has removed this upgrade hook, maybe he'd have a suggestion as to how to address this?

robimarko commented:
OK, as I only moved the check for initramfs into the regular platform_do_upgrade()

spolack commented:
So i understand this that you have some clue what could be causing this?


> Depends on how you "rebased". the hap-ac2 PR and this PR have conflicting changes (4K_SECTORS, pre_upgrade()) which can explain your symptom.

Basically i just took @robimarko branch and cherrypicked @rogerpueyo final commit onto that branch. Mostly it applied fine, but there were some minor things i had to adopt  manually like reorganisation of mikrotik devices into subtarget makefile or upgrade script.

Am i missing something to adjust in DTS, as you are refering to 4K_SECTORS?


If not, could be follwing PR be related?
https://github.com/openwrt/openwrt/pull/3616

timprepscius commented:
My router finally came.  I'd like to test out openwrt on it.

@spolack

How did you build this branch?
I assume you, 
1. clone this the rogerpueyo:ipq40xx-mikrotik-sxtsq-5-ac git
2. run the 1&2&3&4 from Quickstart

On the makemenu
I do not see the MikroTik SXTsq 5 ac listed anywhere.  Should I?

Should I take this conversation elsewhere?  Is it ok to have on this thread?

PolynomialDivision commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -620,6 +620,14 @@ define Device/mikrotik_hap-ac2
+	$(call Device/MikrotikNOR)
+	DEVICE_MODEL := SXTsq 5 ac (RBSXTsqG-5acD)
+	SOC := qcom-ipq4018
+	DEVICE_PACKAGES := rssileds ipq-wifi-mikrotik_sxtsq-5-ac
```
Why not `-kmod-ath10k-ct kmod-ath10k-ct-smallbuffers`?

PolynomialDivision commented:
BTW: I rebased the latest changes:
https://github.com/PolynomialDivision/openwrt/tree/rebase-sxtsq-5-ac

timprepscius commented:
Hey there, I have a question, but maybe this is the inappropriate place to ask, if so, my apologies.

So I've generated I've built based off of this branch.   And I have succesfully done a test run of the initramfs it generated.
But there is no http/luci in the initramfs.  And I'd like to test that as well.

Is there a way to build it so that the initramfs includes this pacakge?

rogerpueyo commented:
> Why not -kmod-ath10k-ct kmod-ath10k-ct-smallbuffers?

The device has plenty of RAM, so there's no need to use the smallbuffers variant of ath10k-ct.

> Hey there, I have a question, but maybe this is the inappropriate place to ask, if so, my apologies.

There's the explanation on how to do it right here https://github.com/openwrt/openwrt/pull/3108#issuecomment-742177483 , you just have to toggle ``LuCi => 1. Collections => <*> luci`` (not ``<m> luci``). Or just use the attached [configuration file](https://github.com/openwrt/openwrt/files/5711111/config.txt) (rename it to .config).

> BTW: I rebased the latest changes:
> https://github.com/PolynomialDivision/openwrt/tree/rebase-sxtsq-5-ac

Did you rebase from master or from the hAP AC² support pull request?

PolynomialDivision commented:
> The device has plenty of RAM, so there's no need to use the smallbuffers variant of ath10k-ct.

Yep. Sry, I did not notice that the default target already has all kernel modules selected.

> Did you rebase from master or from the hAP AC² support pull request?

I did robin marko, rebased and cherry-picked from you with small edits.

john-tho reviewed (commented):
Hi @rogerpueyo, I have a few suggestions from a quick look over the `add support for MikroTik SXTsq 5 ac` commit.

Thanks.

rogerpueyo commented on target/linux/ipq40xx/base-files/etc/board.d/01_leds:
```diff
@@ -46,6 +46,14 @@ engenius,ens620ext)
 	;;
+mikrotik,sxtsq-5-ac)
+	ucidef_set_rssimon "wlan0" "200000" "1"
+	ucidef_set_led_rssi "rssilow" "rssilow" "${boardname}:green:rssilow" "wlan0" "1" "100"
+	ucidef_set_led_rssi "rssimediumlow" "rssimediumlow" "${boardname}:green:rssimediumlow" "wlan0" "21" "100"
+	ucidef_set_led_rssi "rssimedium" "rssimedium" "${boardname}:green:rssimedium" "wlan0" "41" "100"
+	ucidef_set_led_rssi "rssimediumhigh" "rssimediumhigh" "${boardname}:green:rssimediumhigh" "wlan0" "61" "100"
+	ucidef_set_led_rssi "rssihigh" "rssihigh" "${boardname}:green:rssihigh" "wlan0" "81" "100"
```
Done!

rogerpueyo replied:
Done!

rogerpueyo commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-sxtsq-5-ac.dts:
```diff
@@ -0,0 +1,240 @@
+		};
+
+		rssihigh {
+			label = "sxtsq-5-ac:green:rssihigh";
```
Done!

rogerpueyo commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -132,7 +132,8 @@ case "$FIRMWARE" in
 		;;
-	mikrotik,hap-ac2)
+	mikrotik,hap-ac2|\
+	mikrotik,sxtsq-5-ac)
```
My bad, I think I added it in one of these recent rebases.

rogerpueyo commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-sxtsq-5-ac.dts:
```diff
@@ -0,0 +1,240 @@
+				#address-cells = <1>;
+				#size-cells = <1>;
+				label = "RouterBoot-partitions";
+				reg = <0x0 0x100000>;
```
I originally based the DTS on the hap ac²'s, which might have changed since then, but I haven't touched mine.

I've cloned the `&blsp1_spi1` node from hap ac².

rogerpueyo commented:
Hi @john-tho,

Thank you very much for the review! I've made the suggested changes and force-pushed the new commit.

> Once hapac2 is stable/merged, we should use a common dtsi for these ipq4018 mikrotik NOR devices?

Definitely. The two devices have quite many things in common, and possibly more ipq4018 MikroTik NOR devices will come. So let's wait for the hap ac² to be merged.

spolack commented:
I can confirm with latest commit cherry-picked on top of @robimarko PR it works well. sysupgrade works also as expected.

Something i noticed is that writing to rootfs is incredibly slow. If i recall correctly about 500KB/s, when upload sysupgrade i now prefer uploading to /tmp which gives me more than 10MB/s

Thank you for your work

f00b4r0 commented:
@spolack can you paste the output of `cat /proc/mtd`

f00b4r0 commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -160,6 +162,15 @@ ipq40xx_setup_macs()
 		wan_mac=$(mtd_get_mac_ascii devinfo hw_mac_addr)
 		lan_mac=$(macaddr_add "$wan_mac" 1)
 		;;
+	mikrotik,hap-ac2)
+		wan_mac=$(cat /sys/firmware/mikrotik/hard_config/mac_base)
+		lan_mac=$(macaddr_add $wan_mac 1)
+		label_mac="$wan_mac"
+		;;
+	mikrotik,sxtsq-5-ac)
+		lan_mac=$(cat /sys/firmware/mikrotik/hard_config/mac_base)
+		label_mac="$lan_mac"
+		;;
```
If this were part of a subtarget-only file (as I suggested on the hapac2 PR), it would be possible to define mac_base for all devices and use it for label_Mac for all devices as well. I suspect keeping this in the main target will result in lots of duplicated segments. My 2c.
@robimarko

robimarko replied:
I partly agree, but having MikroTik only base-files would incur a lot of duplicates currently.
There are simply not enough MikroTik devices in it to justify duplicating everything.

Also, we can simply define the MAC base once.

f00b4r0 replied:
> I partly agree, but having MikroTik only base-files would incur a lot of duplicates currently.
> There are simply not enough MikroTik devices in it to justify duplicating everything.

There are simply not enough MikroTik devices _now_. It might be useful to anticipate a bit :)

f00b4r0 commented on target/linux/ipq40xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -110,6 +110,11 @@ platform_do_upgrade() {
 		CI_KERNPART="part.safe"
 		nand_do_upgrade "$1"
 		;;
+	mikrotik,hap-ac2|\
+	mikrotik,sxtsq-5-ac)
+		[ "$(rootfs_type)" = "tmpfs" ] && mtd erase firmware
+		default_do_upgrade "$1"
+		;;
```
Likewise here we will end up enumerating all devices instead of having a catchall. Maybe this could be mitigated with `mikrotik,*)`.
@robimarko

robimarko replied:
Not that won't work for all of the devices like this only works for SPI-NOR based ones, there are also NAND based ones like hAP ac3 for which this does not work.

Honestly, I don't see an issue with this.

f00b4r0 replied:
You can special case the NAND devices before the catchall.
That being said, I will reiterate: we should NOT add support for more NAND devices. We are deceiving users and we will brick their devices in the end.

robimarko replied:
Yeah, we can do that.
I personally prefer listing every device as it makes it really clear, especially with the MikroTik revision mess.

Why not?
IPQ40xx based NAND ones like RB450Gx4 are not like the MIPS ones.
These require the kernel to be packaged in UBIFS partition while rootfs uses the UBI wrapper, so bad blocks are actually managed normally.

f00b4r0 replied:
> Yeah, we can do that.
> I personally prefer listing every device as it makes it really clear, especially with the MikroTik revision mess.

It also wastes space and makes script execution slower...

> Why not?
> IPQ40xx based NAND ones like RB450Gx4 are not like the MIPS ones.
> These require the kernel to be packaged in UBIFS partition while rootfs uses the UBI wrapper, so bad blocks are actually managed normally.

That's good news. But then I don't understand why the NOR devices still use YAFFS if mikrotik is already switching to UBIFS on NAND. Surely they realized that YAFFS is going the way of the dinosaurs? Maybe it's possible to get rid of kernel2minor on the ipq devices eventually?

adschm replied:
> You can special case the NAND devices before the catchall.
> That being said, I will reiterate: we should NOT add support for more NAND devices. We are deceiving users and we will brick their devices in the end.

Maybe off-topic here, but with the next stable close, I would expect that the issues won't be fixed in time anymore.

I wonder whether we should put `DEFAULT := n` for the relevant Mikrotik devices (just NAND or even all of them). Opinions?

f00b4r0 reviewed (commented):
I still feel we're not properly taking advantage of the subtarget, which I'm afraid will result in a lose-lose situation for both the main and sub.
I second @john-tho's suggestion that a DTSI should be made.
Cheers

PolynomialDivision commented:
Vlans do not work anymore with current master. The device is able to receive packages, but can not send any.
I suspect it has something to do with: https://github.com/openwrt/openwrt/commit/cdc8d4b46f0811292fffc6094acd81e854b5f4a5

@blocktrron any idea?

PolynomialDivision commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-sxtsq-5-ac.dts:
```diff
@@ -0,0 +1,246 @@
+
+		edma@c080000 {
+			qcom,num_gmac = <1>;
+			status = "okay";
```
Could it be that `qcom,single-phy;` is missing here?

rogerpueyo replied:
Adding `qcom,single-phy` makes the `eth0` interface not to be loaded, with this kernel error:
```
[    1.691062] [<c0626ba8>] (driver_register) from [<c0302764>] (do_one_initcall+0x84/0x1d8)
[    1.699305] [<c0302764>] (do_one_initcall) from [<c0b00fcc>] (kernel_init_freeable+0x1c4/0x26c)
[    1.707381] [<c0b00fcc>] (kernel_init_freeable) from [<c086941c>] (kernel_init+0x8/0x114)
[    1.715884] [<c086941c>] (kernel_init) from [<c03010e8>] (ret_from_fork+0x14/0x2c)
[    1.724211] Exception stack(0xcf82bfb0 to 0xcf82bff8)
[    1.731677] bfa0:                                     00000000 00000000 00000000 00000000
[    1.736805] bfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    1.744961] bfe0: 00000000 00000000 00000000 00000000 00000013 00000000
[    1.753161] ---[ end trace eca84138885b5897 ]---
[    1.759814] ------------[ cut here ]------------
[    1.764411] WARNING: CPU: 2 PID: 1 at mm/vmalloc.c:2244 __vunmap+0x25c/0x260
[    1.769002] Trying to vfree() nonexistent vm area ((ptrval))
[    1.776090] Modules linked in:
[    1.781697] CPU: 2 PID: 1 Comm: swapper/0 Tainted: G        W         5.4.89 #0
[    1.784539] Hardware name: Generic DT based system
[    1.791758] [<c030e544>] (unwind_backtrace) from [<c030afb4>] (show_stack+0x10/0x14)
[    1.796618] [<c030afb4>] (show_stack) from [<c08536b8>] (dump_stack+0x94/0xa8)
[    1.804514] [<c08536b8>] (dump_stack) from [<c03213ac>] (__warn+0xbc/0xd8)
[    1.811541] [<c03213ac>] (__warn) from [<c032144c>] (warn_slowpath_fmt+0x84/0x94)
[    1.818402] [<c032144c>] (warn_slowpath_fmt) from [<c041bcdc>] (__vunmap+0x25c/0x260)
[    1.825957] [<c041bcdc>] (__vunmap) from [<c06299c4>] (release_nodes+0x1ac/0x1d8)
[    1.833769] [<c06299c4>] (release_nodes) from [<c0625f98>] (really_probe+0x114/0x358)
[    1.841234] [<c0625f98>] (really_probe) from [<c06265c0>] (device_driver_attach+0x58/0x60)
[    1.849046] [<c06265c0>] (device_driver_attach) from [<c0626620>] (__driver_attach+0x58/0xcc)
[    1.857205] [<c0626620>] (__driver_attach) from [<c06245c8>] (bus_for_each_dev+0x68/0x8c)
[    1.865800] [<c06245c8>] (bus_for_each_dev) from [<c062554c>] (bus_add_driver+0x1c8/0x1d8)
[    1.873960] [<c062554c>] (bus_add_driver) from [<c0626ba8>] (driver_register+0x74/0x108)
[    1.882120] [<c0626ba8>] (driver_register) from [<c0302764>] (do_one_initcall+0x84/0x1d8)
[    1.890367] [<c0302764>] (do_one_initcall) from [<c0b00fcc>] (kernel_init_freeable+0x1c4/0x26c)
[    1.898440] [<c0b00fcc>] (kernel_init_freeable) from [<c086941c>] (kernel_init+0x8/0x114)
[    1.906947] [<c086941c>] (kernel_init) from [<c03010e8>] (ret_from_fork+0x14/0x2c)
[    1.915270] Exception stack(0xcf82bfb0 to 0xcf82bff8)
[    1.922735] bfa0:                                     00000000 00000000 00000000 00000000
[    1.927863] bfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    1.936021] bfe0: 00000000 00000000 00000000 00000000 00000013 00000000
[    1.944211] ---[ end trace eca84138885b5898 ]---
[    1.950640] ess_edma: probe of c080000.edma failed with error -2
```
whereas, without it, it was correctly loaded:
```
[    2.852076] ess_edma c080000.edma eth0: Link is Down
[    6.011935] ess_edma c080000.edma eth0: Link is Up - 100Mbps/Full - flow control off
[    6.012879] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[    7.088016] ess_edma c080000.edma eth0: Link is Down
```


Therefore, either `qcom,single-phy` is not missing, or it must be accompanied by something else...

PolynomialDivision replied:
Whups sry. Yeah I just guessed why vlans are not working. As @spolack alredy mentioned has the device the switch enabled. His workarounds are working. :)

rogerpueyo replied:
No worries. There are a few devices in the ipq40xx target with the feature. I tried using their configurations on the SXT and the result was the same: no eth0 and kernel error. :disappointed:

spolack commented:
> @spolack can you paste the output of `cat /proc/mtd`

`root@********:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00080000 00010000 "Qualcomm"
mtd1: 00080000 00010000 "RouterBoot"
mtd2: 00001000 00010000 "hard_config"
mtd3: 00007c0c 00010000 "dtb_config"
mtd4: 00001000 00010000 "soft_config"
mtd5: 00f00000 00010000 "firmware"
mtd6: 002c0000 00010000 "kernel"
mtd7: 00c40000 00010000 "rootfs"
mtd8: 00740000 00010000 "rootfs_data"
`

f00b4r0 commented:
@spolack this looks correct. BTW you should always upload sysupgrade to /tmp, and never to the overlay. You're wearing the flash needlessly.

spolack commented:
> Vlans do not work anymore with current master. The device is able to receive packages, but can not send any.
@PolynomialDivision 
I can confirm that. Not sure what it is related to as it used to work with the image i've built months ago. But since ~14 days its broken.

lynxis commented:
Please rebase against the current master.

rogerpueyo commented:
> Please rebase against the current master.

_Voilà_. Only compile-tested.

f00b4r0 commented:
A couple review comments are still pending AFAICT

spolack commented:
> > Vlans do not work anymore with current master. The device is able to receive packages, but can not send any.
> > @PolynomialDivision
> > I can confirm that. Not sure what it is related to as it used to work with the image i've built months ago. But since ~14 days its broken.

Found out why VLANs are not working. Device (even though single port) has an internal switch, which is hard wired as usual to VLAN1 (LAN) and VLAN2 (WAN).

Unfortunately this is hidden from `/etc/config/network`.

Problem can be solved by making the internal switch act as an dumb switch:
```config switch
	option name 'switch0'
	option reset '1'
	option enable_vlan '0'
```

Otherwise its possible to use VLANs if you tag them on port 0 and 5.

--> Is it possible to hide the internal switch from the user and disabling its functionality?

rogerpueyo commented:
I hadn't seen this before. You're right:, Simon.
```
root@OpenWrt:~# swconfig dev switch0 show | tail
TxLateCol   : 0

	pvid: 1
	link: port:5 link:up speed:100baseT full-duplex auto
VLAN 1:
	vid: 1
	ports: 0t 5 
VLAN 2:
	vid: 2
	ports: 0t 
```

This is the typical setup for those devices with 2 or more ports, where eth0.1 is used for LAN and eth0.2 for WAN. However, this is not our case. As @PolynomialDivision spotted, it might be related to commit [cdc8d4b](https://github.com/openwrt/openwrt/commit/cdc8d4b46f0811292fffc6094acd81e854b5f4a5).

There is a [comment at the end](https://github.com/openwrt/openwrt/commit/cdc8d4b46f0811292fffc6094acd81e854b5f4a5#commitcomment-45278622) of the above-mentioned commit, by NoTengoBattery, with a couple of patches linked. By reading the description of the first patch I understand that this VLAN tagging is the driver default, to take advantage of the hardware and improve network performance on devices with two ports (i.e., not our case).

@spolack Does reverting commit [cdc8d4b](https://github.com/openwrt/openwrt/commit/cdc8d4b46f0811292fffc6094acd81e854b5f4a5) fix the VLANs issue for you?

> --> Is it possible to hide the internal switch from the user and disabling its functionality?

I assume yes, but I understand we should find a way to disable at the driver level this LAN-1 WAN-2 default tagging, which is beyond my knowledge. :disappointed:

PolynomialDivision commented:
Anything new? This device is a worthy opponent to ubiquiti. ;)

PolynomialDivision commented:
@rogerpueyo Couldn't we just use
```
ucidef_add_switch "switch0" ...
```

PolynomialDivision commented:
We could also just extend the `ucidef_add_switch` function by an "enable" flag, so we can set the switch, and directly disable it by default. Would require treewide change of this function... Not sure if this would get accepted...

PolynomialDivision commented:
@rogerpueyo We could just add 

```
ucidef_add_switch "switch0"
ucidef_add_switch_attr "switch0" "enable" "false"
```

spolack commented:
> @rogerpueyo We could just add
> 
> ```
I guess you mean

> ucidef_add_switch "switch0"
> ucidef_add_switch_attr "switch0" "enable_vlans" "false"
> ```

PolynomialDivision commented:
@rogerpueyo The other device is merged! :D
 https://github.com/openwrt/openwrt/commit/faea7becafc634dc882f0ad8f9388862ccd5bd0d
 
 I think you should rebase on current master.

rogerpueyo commented:
> @rogerpueyo The other device is merged! :D
> [faea7be](https://github.com/openwrt/openwrt/commit/faea7becafc634dc882f0ad8f9388862ccd5bd0d)
> 
> I think you should rebase on current master.

Cool! I'll be doing it during this week :-)

rogerpueyo commented:
I rebased current master and force-pushed. **Only compile-tested.**

john-tho commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-sxtsq-5-ac.dts:
```diff
@@ -0,0 +1,246 @@
+				};
+			};
+
+			partition@1000000 {
```
One too many zeroes in the node name.
Picked up in hap ac2 here: https://github.com/openwrt/openwrt/pull/3037#discussion_r573224727

rogerpueyo replied:
Thanks, I didn't see this one! :open_mouth:

rogerpueyo commented:
> I guess you mean
> 
> ```
> ucidef_add_switch "switch0"
> ucidef_add_switch_attr "switch0" "enable_vlans" "false"
> ```

I don't see this anywhere for other devices. Maybe:
```
	mikrotik,sxtsq-5-ac)
		ucidef_add_switch "switch0"
		ucidef_add_switch_attr "switch0" "enable" "false"
		ucidef_set_interface_lan "eth0"
		;;
```
?

PolynomialDivision commented:
> I don't see this anywhere for other devices. If I add it to 02_network, I get:

Looks good! :) I wouid prefer to add something like this.
But I do not care. I just hope that this SoC finally gets into OpenWrt. ^^

spolack commented:
last time i tried it didnt work that way. so i would prefer to configure a switch with 2 ports.

rogerpueyo commented:
> last time i tried it didnt work that way anymore. so i would prefer to configure a switch with 2 ports.
> 
> ```
> 	ucidef_add_switch "switch0" \
> 			"0u@eth0" "5:lan"
> ```

I played a little bit with these settings and the DTS and, after looking at the [Fritz Repeater 1200 DTS](https://github.com/openwrt/openwrt/blob/dc61c32fdc5ec2c47a3857a17a6fb2dbd43c711a/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzrepeater-1200.dts#L63) I gave the RGMII RMII mac mode a try:

```
		// ess-switch is disabled by default as per commit 9deeec35dd3b628b95624e41d4e04acf728991ba
		//		ess-switch@c000000 {
		//			switch_mac_mode = <0x3>; /* mac mode for RGMII RMII */
		//			switch_lan_bmp = <0x0>; /* lan port bitmap */
		//			switch_wan_bmp = <0x10>; /* wan port bitmap */
		//		};


		edma@c080000 {
			status = "okay";
			phy-mode = "rgmii-rxid";
			qcom,num_gmac = <1>;
			qcom,single-phy;
		};
```

This way the switch is not enabled and VLANs can be created as:
```
root@OpenWrt:~# swconfig list
root@OpenWrt:~# tail -n7 /etc/config/network 

config interface 'vlan7'
	option proto 'static'
	option netmask '255.255.255.0'
	option ipaddr '192.168.7.1'
	option ifname 'eth0.7'

root@OpenWrt:~# ip a s eth0.7
16: eth0.7@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether c4:ad:34:fe:fe:fe brd ff:ff:ff:ff:ff:ff
    inet 192.168.7.1/24 brd 192.168.7.255 scope global eth0.7
       valid_lft forever preferred_lft forever
    inet6 fe80::c6ad:34fa:fafa:7fbf/64 scope link 
       valid_lft forever preferred_lft forever
```

spolack commented:
Were you able verify it's functionality? Unfortunately i've currently no test unit.

Just being able to configure that, doesnt mean the switch will actually pass that vlan.

rogerpueyo commented:
> Were you able verify it's functionality? Unfortunately i've currently no test unit.
> 
> Just being able to configure that, doesnt mean the switch will actually pass that vlan.

Yes, I set up VLAN7 on my PC and could ping/SSH/capture packets on both directions with WireShark when both naked eth0 (br-lan) and eth0.7 are configured:
![image](https://user-images.githubusercontent.com/6009603/114301626-fb488200-9ac5-11eb-8546-f462f324f963.png)
![image](https://user-images.githubusercontent.com/6009603/114301628-013e6300-9ac6-11eb-92df-14122831c259.png)

rogerpueyo commented:
@spolack 

After my finding:
> I played a little bit with these settings and the DTS and, after looking at the [Fritz Repeater 1200 DTS](https://github.com/openwrt/openwrt/blob/dc61c32fdc5ec2c47a3857a17a6fb2dbd43c711a/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-fritzrepeater-1200.dts#L63) I gave the RGMII RMII mac mode a try:

I gave all the gigabit phy-modes a try: "rgmii", "rgmii-rxid", "rgmii-txid", "rgmii-id", "gmii, "sgmii" and "qsgmii". Except for "sgmii", which didn't work at 100 Mbps, I got these very similar `iperf3` results for all of them:
```
1000 Mbps: 930/940 Mbps,     0/0 rtx (u/d)
100 Mbps:    94/94 Mbps,     0/0 rtx (u/d)
10 Mbps:   7.5/6.8 Mbps, 772/985 rtx (u/d)
```
100 and 1000 Mbps worked nice in all tests, and 10 Mbps speed always showed some retransmissions (very consistent numbers through all the tests).

I am unsure now which phy-mode is the correct one. :disappointed:

PolynomialDivision commented:
> I am unsure now which phy-mode is the correct one. disappointed

I think it is like that:
"rgmii" should be the preferred mode, which allows the mac layer to turn off the dealys completely if they are not needed.

rogerpueyo commented:
>  "rgmii" should be the preferred mode, which allows the mac layer to turn off the dealys completely if they are not needed.

Nice. I left `rgmii`, since it worked correctly (or so it seems, after a couple of days up) and I didn't know which delays to set. Hopefully everything is ready for merging now.

PolynomialDivision commented:
So we can merge? :)

spolack commented:
@lynxis Is there anything we can do to speed up getting this merged?

timprepscius commented:
Hey there, sorry for butting in on this conversation.

When using this firmware (although from a few months ago) in the wild, I
noticed this progression:

1. WRT Connects to WiFi mschapv2
2. Everything is fine
3. WiFi kicks WRT off by expiring encryption credentials, after session
time expired (they set for an hour)
4. Instead of trying to immediately reauthenticate on same frequency WRT
does a process which takes about 30 second and up to 5 minutes.

Is this anything that this patch could be affecting? Should I try again
with the newer version?

When I’ve used a different WRT router to connect to said WiFi, I did not
notice this problem.

The router board firmware takes 8 seconds, and if I restrict the
frequencies it queries it can take 1.

-Tim

On Wed, Apr 21, 2021 at 8:50 AM Simon Polack ***@***.***>
wrote:

> @lynxis <https://github.com/lynxis> Is there anything we can do to speed
> up getting this merged?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/3108#issuecomment-824034104>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZQXZQHM56SZA745GCG7SDTJ3C2BANCNFSM4N5DA5KQ>
> .
>

john-tho commented:
> 1. WRT Connects to WiFi mschapv2
> 3. WiFi kicks WRT off
> 4. Instead of trying to immediately reauthenticate on same frequency WRT does a process which takes about 30 second and up to 5 minutes.

@timprepscius 
The delay reconnecting: This might be 802.11h / DFS? Make sure you have your country set for the wireless radio on this device. You can check the rules hostapd is using with: `iw reg get`. Current connection details in `iw dev`, and full radio details in `iw phy`. Otherwise, we will need logs: `dmesg` and `logread` from startup, to after when this is happens. If you need help checking or setting these settings, or getting the logs, the forum is the best place to ask: https://forum.openwrt.org/

rogerpueyo commented:
@timprepscius 
I don't think it's related with the patch that adds support to the device itself, but with some aspect of the wireless driver, hostapd/wpad. As John suggested, it could be some issue with DFS (although I understand the device is in client mode). It would be nice to know if other devices with similar hardware (specially, wifi) and running OpenWrt experience the same.

In any case, the forum is the best place to ask. :-)

timprepscius commented:
Thank you for this information,

-tim

On 4/22/21, rogerpueyo ***@***.***> wrote:
> @timprepscius
> I don't think it's related with the patch that adds support to the device
> itself, but with some aspect of the wireless driver, hostapd/wpad. As John
> suggested, it could be some issue with DFS (although I understand the device
> is in client mode). It would be nice to know if other devices with similar
> hardware (specially, wifi) and running OpenWrt experience the same.
>
> In any case, the forum is the best place to ask. :-)
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/openwrt/openwrt/pull/3108#issuecomment-825014269

aparcar commented:
I just merged this. Compiled fine on my machine and multiple users seem to use this without issues, scrolling through the comments and talking to @rogerpueyo I couldn't find any blocking issues.

As always, **thanks** for your contribution Roger.

rogerpueyo commented:
Thanks for merging it, @aparcar. The patch is based on @robimarko's code for ipq40xx-mikrotik, so let's also thanks him. :+1:
